### ISSUE TITLE
我模仿你的布局，headerView下滑的时候，mjRefresh没有下拉刷新，什么原因？

### ISSUE THREAD

**Issue by zhouxinle** [Roylee-ML/SwipeTableView#53]
根据headerView frame的改变，来设置下面tableview的contentOffset，tableview可以跟着header view一起滚动了，下拉刷新mj_header箭头也露出来了，但是没有刷新效果，是不是跟你提到的：这里采用UIKit Dynamic物理动画引擎自定义STHeaderView实现自定义UIScrollView效果解决上述问题  这个有关？ 我没有用这个动画。
求解答，谢谢了

**Comment by zhouxinle** [comment]
![Uploading 271125B7-87E8-476E-B1B7-0879A79F3F6B.png…]()

**Comment by zhouxinle** [comment]
![271125b7-87e8-476e-b1b7-0879a79f3f6b](https://cloud.githubusercontent.com/assets/4567999/24733927/3063cb22-1aad-11e7-9e52-4e8baa2ad69a.png)

**Comment by Roylee-ML** [comment]
MJ 的处理主要是在`- (void)scrollViewContentOffsetDidChange:(NSDictionary *)change`这个方法中，会根据`isDragging`来判断当前的 scrollview 是不是处于拖拽的过程。

如果你是通过手势处理 headerview，你需要在手势结束的时候，主动为当前的 scrollview 设置一次`contentOffset`，来触发 KVO，MJ 才会监听到变化，从而改变 MJ 的`state`来生成刷新状态：

```objc
if (yourGesture.state == .end) {
    CGPoint offset = scrollview.contentOffset;
    // 这里主动设置一次 offset
    scrollview.contentOffset = offset;
}
```

具体的方面你可以看下 MJ 的源码，研究一下他的变化过程。

**Comment by zhouxinle** [comment]
我看了，你好像是在这里判断self.scrollview.isDragging的，求确认。
- (BOOL)st_isDragging {
    SwipeTableView * swipeTableView = self.swipeTableView;
    UIView * headerView = swipeTableView.swipeHeaderView;
    if ([headerView isKindOfClass:STHeaderView.class]) {
        STHeaderView * header = (STHeaderView *)headerView;
        if (header.isDragging && self == swipeTableView.currentItemView) {
            return YES;
        }
        return [self st_isDragging];
    }
    return [self st_isDragging];
}

**Comment by Roylee-ML** [comment]
是的。我是在这里做了hook，这样才能保证 拖拽自定义的 headerview 也是拖拽 scrollview。但是，在手势结束的时候还是要主动 setter 一次 offset，触发 KVO 来触发 MJ 的监听。

**Comment by hilen** [comment]
这里可以用非 hook 的形式来实现吗？

**Comment by ioswukai** [comment]
我的header是SwipeHeaderView 上面放了一个UICollectionView的轮播    手按着上面的header轮播向下滑动 停止滑动了 页面也不返回到上面  如下图 
![858f09adcc2aad729f95bce123b1a827](https://cloud.githubusercontent.com/assets/12831122/26387155/e25a1b1a-407d-11e7-9b30-1c29b28d1403.png)

**Comment by Homin818** [comment]
还是不太懂怎么添加MJRefreshHeader 和 MJRefreshFooter

**Comment by Roylee-ML** [comment]
目前主要维护的是3.x 的分支。建议采用这个分支（由于文档一直没有写，所以没有合到 master），下拉刷新 swipetableview 已经不再干预，正常接入任何三方下拉刷新即可😊
